# Линейное программирование
[Источник](http://personal.vu.nl/l.stougie/Courses/ALP/BTonlyCh12345.pdf)

## Задача 1. Низкий старт
Сформулируйте следующую задачу как задачу линейного программирования (ЛП)
$$
\begin{split}
 \mathrm{min} \;\;\;4 x_1 + 5\left|x_2 - 1\right| &\\
\text{s.t. }  2\left|x_1\right| +\left|x_2-3\right| &\leq 5\\
\end{split} 
$$

__TL;DR _Можно сделать замену модуля, предложенную ниже, но в источнике указана более удачная.___

Постараемся найти наиболее общий подход. Так как $|x| = \max\{x, -x\}$, то также эквивалентна замена:

$$
\begin{cases}
|x| \rightarrow \min w \\
|x| \le w 
\end{cases}
$$

Преобразуем исходную задачу:

$$
\min_{x_1, x_2 }(4x_1 + 5 \min_{\substack{x_3 \\ |x_2 - 1| \le x_3} } x_3)\\
2|x_1| + |x_2 - 3| \le 5
$$

Вынесем минимум:

$$
\min_{\substack{x_1, x_2 \\ 2|x_1| + |x_2 - 3| \le 5} }\min_{\substack{x_3 \\ |x_2 - 1| \le x_3} }(4x_1 + 5 x_3)
$$

Хотелось бы, чтобы можно было переписать так:

$$
\min_{\substack{x_1, x_2, x_3 \\ 2|x_1| + |x_2 - 3| \le 5 \\ |x_2 - 1| \le x_3} }(4x_1 + 5 x_3)
$$


Функция $f(x_1, x_2, x_3) = 4x_1 + 0x_2 + 5 x_3 = 4x_1 + 5 x_3$ -- линейная, значит, достигает минимума на границе. Обозначим множества: $C_1 = \{ x_1, x_2 \; | \;  2|x_1| + |x_2 - 3| \le 5\}$, $C_2 = \{x_3 \; | \; |x_2 - 1| \le x_3\}$ и $C_3 = C_1 \cap C_2$. Тут полезно порисовать. $C_1$ -- цилиндр с осью, параллельной $x_3$, $C_2$ -- долина вдоль оси $x_1$. Тогда функция $f(x_1, x_2, x_3)$ достигает минимума на границе $\partial C_2$. Осталось показать, что $C_1 \cap \partial C_2 \simeq C_3$ в том смысле, что на этих множествах минимальные значения функции совпадают, тогда такое занесение ограничений под общий минимум легально. Заметим, что 
$$\underset{\substack{(x_1, x_2, x_3) \\ |x_2 - 1| \le x_3} } {\textbf{argmin} } x_3 = \partial C_2$$
Но так как $f(x_1, x_2, x_3) =  4x_1 + 5 x_3$, и коэффициент перед $x_3 > 0$ (__важно__, перемещение вниз по $x_3$ уменьшает целевую функцию), то: 
$$(\underset{(x_1, x_2, x_3) \in C_3} {\textbf{argmin} } [4x_1 + 5 x_3]) \in C_3 \cap \partial C_2$$

$$C_3 \cap \partial C_2 = C_1 \cap C_2 \cap \partial C_2 = C_1 \cap \partial C_2$$

То есть показали требуемое. Если бы коэффициент перед $|x_2 - 1|$ оказался отрицательным, то рассмотрели бы замену (__важное__ замечание снова сработает). Однако, теперь задача распадётся на 2 подазадачи (после раскрытия модуля), что хуже:

$$
\begin{cases}
-|x| \rightarrow \min w \\
-|x| \le w 
\end{cases}
$$

$$
\min (4x_1 + 5 x_3) \\
2|x_1| + |x_2 - 3| \le 5 \\
|x_2 - 1| \le x_3
$$

$$
$$

$$
\min (4x_1 + 5 x_3) \\
2x_1  \le 5 - |x_2 - 3| \\
|x_2 - 3| - 5  \le 2x_1  \\
x_2 - 1 \le x_3 \\
-x_3 \le x_2 - 1 \\
$$

$$
$$

$$
\min (4x_1 + 5 x_3) \\
|x_2 - 3| \le 5 - 2x_1  \\
|x_2 - 3|  \le 2x_1 + 5 \\
x_2 - 1 \le x_3 \\
-x_3 \le x_2 - 1 \\
$$

$$
$$

$$
\min (4x_1 + 5 x_3) \\
x_2 - 3 \le 5 - 2x_1  \\
2x_1 - 5 \le x_2 - 3 \\ 
x_2 - 3 \le 2x_1 + 5 \\
-2x_1 - 5 \le x_2 - 3 \\ 
x_2 - 1 \le x_3 \\
-x_3 \le x_2 - 1 \\
$$

$$
$$

$$
\min (4x_1 + 5 x_3) \\
2x_1 + x_2  \le 8 \\
2x_1 - x_2  \le  2 \\ 
-2x_1 + x_2 \le 8 \\
-2x_1 - x_2 \le 2 \\ 
x_2 -x_3  \le 1 \\
-x_2 -x_3 \le -1 \\
$$

$$
$$

$$
\min \begin{bmatrix}
4 & 0 & 5 
\end{bmatrix}
\begin{bmatrix}
x_1 & x_2 & x_3 
\end{bmatrix}^T\\
\begin{bmatrix}
2 & 1 & 0 \\
2 & -1 & 0 \\
-2 & 1 & 0 \\
-2 & -1 & 0 \\
0 & 1 & -1 \\
0 & -1 & -1 \\
\end{bmatrix}
\begin{bmatrix}
x_1 \\
x_2 \\
x_3
\end{bmatrix}
\le
\begin{bmatrix}
8 \\
2 \\
8 \\
2 \\ 
1 \\
-1
\end{bmatrix}
$$

## Задача 2. Fuck the tables
Рассмотрите задачу 
$$
\begin{split}
 \mathrm{min} -x_1 - 2x_2 &\\
\text{s.t. }  -x_1 +3x_2 &\leq 3\\
                  x_1 + x_2 &\leq 5\\
                  x_1, x_2 \geq 0 .
\end{split} 
$$

### 1.
Приведите задачу к стандартной (канонической) форме задачи ЛП и найдите начальное решение (которое можно давать решать симплекс методу), в котором $(x_1, x_2) = (0, 0)$.

$$
\min -x_1 - 2x_2  \\
-x_1 + 3x_2 + x_3 = 3 \\
x_1 + x_2 + x_4 = 5 \\
x_1, x_2, x_3, x_4 \ge 0
$$

$$
$$

$$
\min 
\begin{bmatrix}
-1 & -2 & 0 & 0
\end{bmatrix}
\begin{bmatrix}
x_1 & x_2 & x_3 & x_4
\end{bmatrix}^T \\
\begin{bmatrix}
-1 & 3 & 1 & 0 \\
1 & 1 & 0 & 1 
\end{bmatrix}
\begin{bmatrix}
x_1 \\
x_2 \\
x_3 \\
x_4
\end{bmatrix}
=
\begin{bmatrix}
3 \\
5 
\end{bmatrix} \\
x_1, x_2, x_3, x_4 \ge 0
$$

Начальное решение:
$$
\begin{bmatrix}
x_1 \\
x_2 \\
x_3 \\
x_4
\end{bmatrix}
=
\begin{bmatrix}
0 \\
0 \\
3 \\
5
\end{bmatrix}
$$

### 2.
Решите задачу в канонической форме с помощью любой библиотеки решения ЛП для python симплекс методом. Выводите при этом нашу любимую таблицу решений.

In [1]:
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import linprog
from matplotlib.animation import FuncAnimation
plt.rcParams["animation.html"] = "html5"
np.set_printoptions(suppress=True, precision=2)

In [2]:
c = np.array([-1, -2, 0, 0])
A_eq = np.array([[-1, 3, 1, 0],
                 [ 1, 1, 0, 1]])
b_eq = np.array([3, 5])

def cb(xk, **kwargs):
    print(kwargs['tableau'], '\n')

res = linprog(c, A_eq=A_eq, b_eq=b_eq, callback=cb)
print(res.message)
print('f = %g' % res.fun)
print('x = %s' % res.x)

[[-1.  3.  1.  0.  1.  0.  3.]
 [ 1.  1.  0.  1.  0.  1.  5.]
 [-1. -2.  0.  0.  0.  0.  0.]
 [ 0. -4. -1. -1.  0.  0. -8.]] 

[[-0.33  1.    0.33  0.    0.33  0.    1.  ]
 [ 1.33  0.   -0.33  1.   -0.33  1.    4.  ]
 [-1.67  0.    0.67  0.    0.67  0.    2.  ]
 [-1.33  0.    0.33 -1.    1.33  0.   -4.  ]] 

[[ 0.    1.    0.25  0.25  0.25  0.25  2.  ]
 [ 1.    0.   -0.25  0.75 -0.25  0.75  3.  ]
 [ 0.    0.    0.25  1.25  0.25  1.25  7.  ]
 [ 0.    0.   -0.    0.    1.    1.    0.  ]] 

[[ 0.    1.    0.25  0.25  2.  ]
 [ 1.    0.   -0.25  0.75  3.  ]
 [ 0.    0.    0.25  1.25  7.  ]] 

Optimization terminated successfully.
f = -7
x = [3. 2. 0. 0.]


### 3.
Визуализируйте задачу в исходной формулировке и отрисуйте на ней шаги симплекс метода.

_Сразу находится оптимальное решение. Добавил дополнительное ограничение, чтобы было не так тоскливо._

$$
0.2 x_1 + x_2 \le 1.5
$$

In [3]:
%%capture --no-stderr --no-stdout
fig, ax = plt.subplots(1, 1, figsize=(8, 5))

nframes = 10

def constraints(x_1, x_2):
    c1 = -x_1 + 3*x_2 < 3
    c2 = x_1 + x_2 <= 5
    c3 = 0.2*x_1 + x_2 <= 1.5
    c4 = x_1 >= 0
    c5 = x_2 >= 0
    return np.all((c1, c2, c3, c4, c5), axis=0)

xs = np.linspace(-0.5, 6, 1000)
ys = np.linspace(-0.5, 3, 1000)

X, Y = np.meshgrid(xs, ys)
Z = constraints(X, Y)
F = -X - 2*Y
ax.contourf(X, Y, Z)
c = ax.contour(X, Y, F)
plt.colorbar(c, ax=ax)

ax.set_title('Simplex method steps')
ax.set_xlabel(r"$x_1$")
ax.set_ylabel(r"$x_2$")

ps = []
def cb(xk, **kwargs):
    ps.append(xk[:2])


c = np.array([-1, -2, 0, 0, 0])
A_eq = np.array([[-1,    3, 1, 0, 0],
                 [ 1,    1, 0, 1, 0],
                 [ 0.2,  1, 0, 0, 1]])
b_eq = np.array([3, 5, 1.5])
    
res = linprog(c, A_eq=A_eq, b_eq=b_eq, callback=cb)
print(res)
ps = np.array(ps)
traj, = ax.plot([], [], 'r--', marker='o')

def init():
    return traj,

def update(frame):
    traj.set_data(ps[:frame, 0], ps[:frame, 1])
    return traj,

anim = FuncAnimation(fig, update, frames=nframes, interval=120,
                   init_func=init, blit=True)
plt.show()

     fun: -5.625
 message: 'Optimization terminated successfully.'
     nit: 4
   slack: array([], dtype=float64)
  status: 0
 success: True
       x: array([4.38, 0.62, 5.5 , 0.  , 0.  ])


In [4]:
anim

## Задача 3. Дорога в светлое будущее
Рассмотрите дорогу из темного прошлого в светлое будущее, разделенную на $n$ участков и освещаемую $m$ лампами. Пусть $j$-ая лампа имеет мощность $p_j$. Освещенность $i$-ого участка дороги определяется как $I_i = \sum\limits_{j=1}^m a_{ij}p_j$, где $a_{ij}$ - известные коэффициенты. Мы хотим, чтобы $i$-ый участок дороги имел освещенность $I^*_i$.

Наша задача состоит в том, чтобы подобрать мощности ламп $p_j$ так, чтобы освещенности $I_i$ всех участков дороги были близки к желаемым $I_i^*$.

Составьте и решите оптимизационную задачу (любым способом) для оптимального освещения дороги в форме задачи ЛП (обратите внимание, что формулировок может быть много - выберите и объясните почему Вам понравилась именно эта).

Примечание: $a, I^*$ генерируйте случайно.

Сформулируем задачу в виде проецирования (в смысле $l_1$ нормы) оптимального вектора $[I_1^*, \dots, I_n^*]$ на допустимое множество $[I_1, \dots, I_n]$, $I_i = \sum_{j=1}^m a_{ij}p_j$, $p_j \ge 0$. Тогда запишем:

$$
\min ||Ap-I^*||_{l_1} \\
p_i \ge 0, \; i = \overline{1, m}
$$

$$
$$

$$
\min \; \sum_{i=1}^n\left|\sum_{j=1}^m a_{ij}p_j - I^*\right| \\
p_i \ge 0, \; i = \overline{1, m}
$$

Теперь раскроем модуль с помощью приёма из _источника_:

$$
\min \; \sum_{i=1}^n (x_i^+ + x_i^-)\\
\sum_{j=1}^m a_{ij}p_j - I^* = x_i^+ - x_i^-, \; i = \overline{1, n} \\
p_i \ge 0, \; i = \overline{1, m}\\
x_i^+ \ge 0, \; i = \overline{1, n}\\
x_i^- \ge 0, \; i = \overline{1, n}
$$

$$
$$

$$
\min \; \sum_{i=1}^n (x_i^+ + x_i^-)\\
\sum_{j=1}^m a_{ij}p_j - x_i^+ + x_i^- = I^*, \; i = \overline{1, n} \\
p_i \ge 0, \; i = \overline{1, m}\\
x_i^+ \ge 0, \; i = \overline{1, n}\\
x_i^- \ge 0, \; i = \overline{1, n}
$$

Осталось только собрать это всё в матрицу.

In [5]:
# Number of segments
n = 10
# Number of lights
m = 5

A = np.random.rand(n, m)
I = np.random.rand(n)

c = np.append(np.zeros(m), np.ones(2*n))
A_eq = np.zeros((n, m + 2*n))
for i in range(n):
    A_eq[i, m+2*i:m+2*(i+1)] = [-1, 1]
A_eq[:, :m] = A
b_eq = I
    
res = linprog(c, A_eq=A_eq, b_eq=b_eq)
print(res.message)
print('d = %g' % res.fun)
print('p = %s' % res.x[:m])

Optimization terminated successfully.
d = 1.73351
p = [0.37 0.   0.3  0.   0.37]


## Задача 4. Я у мамы бизнесмен
Вы - начинающий предприниматель и производите 2 вида продуктов. I тип продукта требует $1/4$ часа на сборку, $1/8$ часа на тестирование, а стоимость материалов для его производства равна $1.2$ биткоина. Продукт II типа требует $1/3$ часа на сборку, $1/3$ часа на тестирование, но стоимость материалов его производства равна $0.9$ биткоинов. При текущем штате сотрудников в день возможны лишь $90$ часов сборки и  $80$ часов тестирования.

Стоимость продукта I типа на рынке: $9$ биткоинов, а II типа - $8$.

### 1. 
Сформулируйте задачу ЛП, которая может быть применена здесь для максимизации ежедневной прибыли компании. Найдите сколько оптимально производить продуктов каждого из типов.

Тип продукта | Время на сборку, ч | Время на тестирование, ч | Стоимость материалов, бк | Стоимость на рынке, бк
-------------|--------------------|--------------------------|--------------------------|-----------------------
I            | 1/4                |  1/8                     |   1.2                    | 9
II           | 1/3                |  1/3                     |   0.9                    | 8
__Доступно__ | 90                 |  80                      |                          |


$$
\min -((9-1.2)x + (8-0.9)y) \\
\frac{1}{4}x + \frac{1}{3}y \le 90 \\
\frac{1}{8}x + \frac{1}{3}y \le 80 \\
x, y \ge 0
$$

$$
$$

$$
\min -7.8x - 7.1y \\
\frac{1}{4}x + \frac{1}{3}y \le 90 \\
\frac{1}{8}x + \frac{1}{3}y \le 80 \\
x, y \ge 0
$$

_Эвристика._ Если решения окажутся нецелыми, выберем оптимальное, округляя полученные нецелые в ту или иную сторону. 

In [6]:
c = np.array([-7.8, -7.1])
A = np.array([[1/4, 1/3],
              [1/8, 1/3]])
b = np.array([90, 80])

res = linprog(c, A_ub=A, b_ub=b)
print(res.message)
print('Profit: %g bc' % -res.fun)
print('[Type I, Type II] = %s' % res.x)

Optimization terminated successfully.
Profit: 2808 bc
[Type I, Type II] = [360.   0.]


Максимальная прибыль равна $2808$bc при производстве $360$ изделий типа I и $0$ изделий типа II.

### 2.
Рассмотрите так же две следующие модификации задачи:
#### а) 
В день Вы можете оплачивать до $50$ часов переработки сборщиков за $7$ биткоинов в час
#### б)
Поставщик материалов готов предоставить Вам скидку $10\%$, если Вы закажете в день больше, чем на $300$ биткоинов.

Могут ли такие модификации быть обернуты в задачу ЛП, если да, то как, если нет, то почему, и как бы Вы решали задачу в таком случае?

С возможностью переработки:

$$
\min -7.8x - 7.1y + 7z \\
\frac{1}{4}x + \frac{1}{3}y \le 90 + z \\
\frac{1}{8}x + \frac{1}{3}y \le 80 \\
z \le 50 \\
x, y, z \ge 0
$$

$$
$$

$$
\min -7.8x - 7.1y + 7z \\
\frac{1}{4}x + \frac{1}{3}y - z \le 90 \\
\frac{1}{8}x + \frac{1}{3}y \le 80 \\
z \le 50 \\
x, y, z \ge 0
$$

In [7]:
c = np.array([-7.8, -7.1, 7])
A = np.array([[1/4, 1/3, -1],
              [1/8, 1/3,  0],
              [0  ,   0,  1]])
b = np.array([90, 80, 50])

res = linprog(c, A_ub=A, b_ub=b)
print(res.message)
print('Profit: %g bc' % -res.fun)
print('[Type I, Type II, Overtime] = %s' % res.x)

Optimization terminated successfully.
Profit: 4018 bc
[Type I, Type II, Overtime] = [560.   0.  50.]


Максимальная прибыль равна $4018$bc при производстве $560$ изделий типа I, $0$ изделий типа II и $50$ч переработки.

При условии предоставления скидки задача ставится следующим образом:

$$
\max \max\{(9x + 8y - (1.2x + 0.9y), 9z + 8t -(1.08z + 0.81t)\} \\
\frac{1}{4}x + \frac{1}{3}y \le 90 \\
\frac{1}{8}x + \frac{1}{3}y \le 80 \\
\frac{1}{4}z + \frac{1}{3}t \le 90 \\
\frac{1}{8}z + \frac{1}{3}t \le 80 \\
1.2x + 0.9y \le 300 \\
1.2z + 0.9t \ge 300 \\
x, y, z, t \ge 0
$$

$$
$$

$$
\max \max\{7.8x + 7.1y, 7.92z + 7.19t\} \\
\frac{1}{4}x + \frac{1}{3}y \le 90 \\
\frac{1}{8}x + \frac{1}{3}y \le 80 \\
\frac{1}{4}z + \frac{1}{3}t \le 90 \\
\frac{1}{8}z + \frac{1}{3}t \le 80 \\
1.2x + 0.9y \le 300 \\
1.2z + 0.9t \ge 300\\
x, y, z, t \ge 0
$$
  
  Вот тут возникает проблема, так как ищется максимум нелинейной (ещё и выпуклой) функции. Вопрос "Можно ли свести к задаче ЛП?" достаточно философский, судя по тому, что гугл ответа не знает. Если мы знаем множество, на котором достигается максимум, то задача ЛП ставится тривиально, в этом смысле обернуть можно. Но, не зная ответа, задача сводится к тому, чтобы подобрать подходящую (в том смысле, что максимум должен сохраниться) линейную релаксацию этой задачи. Ясно, что таких бесконечно много, но вот как сконструировать хоть одну лишь по входным данным? 
  
  Так что исходная задача просто подразбивается на 2 задачи, в которых отдельно рассматриваются обе возможности. Затем выбирается лучшее из решений.

$$
\min -7.8x - 7.1y \\
\frac{1}{4}x + \frac{1}{3}y \le 90 \\
\frac{1}{8}x + \frac{1}{3}y \le 80 \\
1.2x + 0.9y \le 300 \\
x, y \ge 0
$$

$$
$$

$$
\min -7.92x - 7.19y \\
\frac{1}{4}x + \frac{1}{3}y \le 90 \\
\frac{1}{8}x + \frac{1}{3}y \le 80 \\
-1.2x - 0.9y \le -300 \\
x, y \ge 0
$$

In [8]:
c1 = np.array([-7.8, -7.1])
A1 = np.array([[1/4, 1/3],
              [1/8, 1/3],
              [1.2, 0.9]])
b1 = np.array([90, 80, 300])

c2 = np.array([-7.92, -7.19])
A2 = np.array([[1/4, 1/3],
              [1/8, 1/3],
              [-1.2, -0.9]])
b2 = np.array([90, 80, -300])

res1 = linprog(c1, A_ub=A1, b_ub=b1)
print('1)', res1.message)
print('   Profit: %g bc' % -res1.fun)
print('   [Type I, Type II] = %s' % res1.x)
res2 = linprog(c2, A_ub=A2, b_ub=b2)
print('2)', res2.message)
print('   Profit: %g bc' % -res2.fun)
print('   [Type I, Type II] = %s' % res2.x)

1) Optimization terminated successfully.
   Profit: 2185.71 bc
   [Type I, Type II] = [108.57 188.57]
2) Optimization terminated successfully.
   Profit: 2851.2 bc
   [Type I, Type II] = [360.   0.]


Сравниваем и делаем вывод, что выгоднее пользоваться скидкой. Тогда максимальная прибыль равна $2851.2$bc при производстве $360$ изделий типа I, $0$ изделий типа II.

__TL;DR _Есть [метод](https://math.stackexchange.com/questions/1858740/how-to-covert-min-min-problem-to-linear-programming-problem/1858751), который позволяет свести эту задачу к ЛП, но с целочисленными ограничениями на некоторые переменные. Он, конечно, не сработает без таких ограничений, но по крайней мере в нашей задаче попутно были найдены оптимальные значения для обоих выражений под максимумами. Совпадение?___

$$
\min \min\{-7.8x - 7.1y, -7.92z - 7.19t\} \\
\frac{1}{4}x + \frac{1}{3}y \le 90 \\
\frac{1}{8}x + \frac{1}{3}y \le 80 \\
\frac{1}{4}z + \frac{1}{3}t \le 90 \\
\frac{1}{8}z + \frac{1}{3}t \le 80 \\
1.2x + 0.9y \le 300 \\
1.2z + 0.9t \ge 300\\
x, y, z, t \ge 0
$$

$$
$$

$$
\min r - p \\
\frac{1}{4}x + \frac{1}{3}y \le 90 \\
\frac{1}{8}x + \frac{1}{3}y \le 80 \\
\frac{1}{4}z + \frac{1}{3}t \le 90 \\
\frac{1}{8}z + \frac{1}{3}t \le 80 \\
1.2x + 0.9y \le 300 \\
-1.2z - 0.9t \le -300\\
-7.8x - 7.1y - Ma \le r - p\\
-7.92z - 7.19t -Mb \le r - p\\
a + b = 1 \\
x, y, z, t, a, b, r \ge 0
$$

In [9]:
# M is large enough
M = 1000000
#             x  y  z  t  a  b  r   p
c = np.array([0, 0, 0, 0, 0, 0, 1, -1])
A = np.array([[ 1/4,   1/3,    0,    0,    0,    0,    0,    0],
              [ 1/8,   1/3,    0,    0,    0,    0,    0,    0],
              [   0,     0,  1/4,  1/3,    0,    0,    0,    0],
              [   0,     0,  1/8,  1/3,    0,    0,    0,    0],
              [ 1.2,   0.9,    0,    0,    0,    0,    0,    0],
              [   0,     0, -1.2, -0.9,    0,    0,    0,    0],
              [-7.8,  -7.1,    0,    0,   -M,    0,   -1,    1],
              [   0,     0,-7.92,-7.19,    0,   -M,   -1,    1]])
b = np.array([90, 80, 90, 80, 300, -300, 0, 0])
A_eq = np.array([[0, 0, 0, 0, 1, 1, 0, 0]])
b_eq = np.array([1])

res = linprog(c, A_ub=A, b_ub=b, A_eq=A_eq, b_eq=b_eq)
print(res.message)
print('Unreal Profit: %g bc' % -res.fun)
print('[x  y  z  t  a  b  r  p] = %s' % res.x)

Optimization terminated successfully.
Unreal Profit: 502518 bc
[x  y  z  t  a  b  r  p] = [   108.57    188.57    360.        0.        0.5       0.5       0.
 502518.46]


Значения $[108.57, 188.57]$ и $[360.,  0.]$ совпадают с оптимальными выше.

In [10]:
from IPython.core.display import HTML
def css_styling():
    styles = open("custom.css", "r").read() #or edit path to custom.css
    return HTML(styles)
css_styling()

FileNotFoundError: [Errno 2] No such file or directory: 'custom.css'